### PR TITLE
`Development`: Optimize local git repository cleanup

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [X] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [X] I **strictly** followed the principle of **data economy** for all database calls.
- [X] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

### Motivation and Context

The two deleted database calls were highly inefficient and took > 30 seconds on production.

I also modified the Gradle run configuration template as currently with every test execution npmInstall and other tasks get executed for no reason

### Description

We actually only need the repository URI for deleting the local repository. Everything else is either irrelevant or only for logging purposes. Thus, I created one query searching both conditions (exercises and exams) at once and only returning the URI.
To have finer control, I reworked it as a paginated run. The batch size is up for debate.

### Steps for Testing

**Local Test only**
Prerequisites:
- Local Programming Setup
- An exam with a programming exercise with the online editor activated
- A course programming exercise with the online editor activated

1. Participate in the exam exercise
2. Participate in the course exercise
3. Make sure that the repository working copies exist in your file system (most likely in the folder `repos`) (be careful if unrelated copies already exist. Don't get confused or delete them beforehand).
4. Make sure the due date of your exercises has passed.
5. Stop your server
6. Go into the `AutomaticProgrammingExerciseCleanupService.cleanupGitWorkingCopiesOnArtemisServer()` and update the `earliestDate` to be before your exercise/exam due dates (i.e. `earliestDate` and `latestDate` should include the due dates).
7. Setup up the `artemis.scheduling.programming-exercises-cleanup-time` to a reasonable schedule for you to test or modify the scheduled parameter of the `cleanup()` method so you don't have to wait.
8. Start the server and wait for the cleanup to commence (watch the logs)
9. The working copies from step 4 are expected to have been removed.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
  - Enhanced cleanup process for Git repositories associated with programming exercises, focusing on local working copies.
  - Introduced new methods for targeted retrieval of programming exercises with eager loading of participations.

- **Bug Fixes**
  - Removed outdated query methods that were no longer necessary, streamlining data access.

- **Chores**
  - Updated Gradle run configuration to exclude additional setup tasks, improving build efficiency.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java [hunk 1]
```diff
@@ -96,35 +102,55 @@ public void cleanupGitRepositoriesOnArtemisServer() {
...
+        log.info("Found {} student participations to clean local student repositories in {} batches.", uriBatch.getTotalElements(), uriBatch.getTotalPages());
+        if (uriBatch.getTotalElements() > 0) {
+            do {
+                uriBatch.forEach(this::deleteLocalRepositoryByUriString);
+                uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate,
+                        uriBatch.nextPageable());
+            }
+            while (uriBatch.hasNext());
+            log.info("Finished cleaning local student repositories");
+        }
```
Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:144]: I am not a big fan of do-while loops. We could use the following code

Comment (krusche) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:144]:
```suggestion
        if (uriBatch.getTotalElements() > 0) {
            Stream<Page<URI>> uriStream = Stream.iterate(uriBatch, batch -> batch.hasNext(),
                batch -> programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, batch.nextPageable()));
    
            uriStream.forEach(batch -> {
                batch.forEach(this::deleteLocalRepositoryByUriString);
            });
        
            log.info("Finished cleaning local student repositories");
        }
```

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java [hunk 2]
```diff
@@ -164,6 +166,19 @@ List<ProgrammingExerciseStudentParticipation> findWithSubmissionsByExerciseIdAnd
...
+                JOIN TREAT (participation.exercise AS ProgrammingExercise) pe
+            WHERE participation.repositoryUri IS NOT NULL
+                AND (
+                    (pe.dueDate IS NOT NULL AND :latestDate <= pe.dueDate AND pe.dueDate <= :earliestDate)
+                    OR (pe.exerciseGroup IS NOT NULL AND :latestDate <= pe.exerciseGroup.exam.endDate AND pe.exerciseGroup.exam.endDate <= :earliestDate)
+                )
+            """)
+    Page<String> findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(@Param("earliestDate") ZonedDateTime earliestDate, @Param("latestDate") ZonedDateTime latestDate,
+            Pageable pageable);
+
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java]: (1) Fixed in 490483bdd56d00a574fe84c870426e74c45a4847
(2) exam is not null implicitly as there never is an exercise group without an exam.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java [hunk 3]
```diff
@@ -82,49 +89,68 @@ public void cleanup() {
...
-                programmingExercises.stream().map(ProgrammingExercise::getProjectKey).collect(Collectors.joining(", ")),
-                programmingExercises.stream().mapToLong(programmingExercise -> programmingExercise.getStudentParticipations().size()).sum());
-        for (var programmingExercise : programmingExercises) {
-            for (var studentParticipation : programmingExercise.getStudentParticipations()) {
-                var programmingExerciseParticipation = (ProgrammingExerciseStudentParticipation) studentParticipation;
-                gitService.deleteLocalRepository(programmingExerciseParticipation.getVcsRepositoryUri());
-            }
-        }
+        var latestDate = earliestDate.minusYears(1).truncatedTo(ChronoUnit.DAYS);
+
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java]: Fixed in 490483bdd56d00a574fe84c870426e74c45a4847

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java [hunk 4]
```diff
@@ -82,49 +89,68 @@
...
+                pageable);
+        log.info("Found {} student participations to clean local student repositories in {} batches.", uriBatch.getTotalElements(), uriBatch.getTotalPages());
+        if (uriBatch.getTotalElements() > 0) {
+            var ignored = Stream.iterate(uriBatch, Page::hasNext, page -> {
+                page.forEach(this::deleteLocalRepositoryByUriString);
+                return programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate,
+                        page.nextPageable());
+            });
+            log.info("Finished cleaning local student repositories");
+        }
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java]: Fixed in 490483bdd56d00a574fe84c870426e74c45a4847

**Review by Strohgelaender — DISMISSED** [review]
code makes sense

File: .idea/runConfigurations/_template__of_Gradle.xml [hunk 5]
```diff
@@ -4,7 +4,7 @@
       <option name="executionName" />
       <option name="externalProjectPath" value="$PROJECT_DIR$" />
       <option name="externalSystemIdString" value="GRADLE" />
-      <option name="scriptParameters" value="-x webapp" />
+      <option name="scriptParameters" value="-x nodeSetup -x npmSetup -x npmInstall -x webapp" />
```
Comment (JohannesStoehr) [inline .idea/runConfigurations/_template__of_Gradle.xml:7]: This should probably not be here

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java [hunk 6]
```diff
@@ -164,6 +166,19 @@ List<ProgrammingExerciseStudentParticipation> findWithSubmissionsByExerciseIdAnd
...
             @Param("username") String username, @Param("testRun") boolean testRun);
 
+    @Query("""
+            SELECT participation.repositoryUri
+            FROM ProgrammingExerciseStudentParticipation participation
+                JOIN TREAT (participation.exercise AS ProgrammingExercise) pe
+            WHERE participation.repositoryUri IS NOT NULL
+                AND (
+                    (pe.dueDate IS NOT NULL AND :earliestDate <= pe.dueDate AND pe.dueDate <= :latestDate)
+                    OR (pe.exerciseGroup IS NOT NULL AND :earliestDate <= pe.exerciseGroup.exam.endDate AND pe.exerciseGroup.exam.endDate <= :latestDate)
```
Comment (magaupp) [inline src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java:176]: JPQL paths create implicit inner joins. Writing `pe.exerciseGroup.exam.endDate` anywhere in the query would result in something like `JOIN pe.exerciseGroup eg JOIN eg.exam e`. This means that if `exerciseGroup` is `NULL` the row is discarded before the `IS NOT NULL` check happens.
Maybe add some `LEFT JOIN`s to get the `exam`.

**Review by magaupp — CHANGES_REQUESTED** [review]
Tested locally with an non-exam exercise with an due date 4 months in the past and multiple participants. The query always returned 0 urls: `Found 0 student participations...`

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java [hunk 7]
```diff
@@ -82,49 +89,66 @@ public void cleanup() {
...
+
+    private void cleanStudentParticipationsRepositories(ZonedDateTime earliestDate, ZonedDateTime latestDate) {
+        log.info("Search for exercises with due date from {} until {}", earliestDate, latestDate);
+        // Get all relevant participation ids
+        Pageable pageable = Pageable.ofSize(STUDENT_PARTICIPATION_CLEANUP_BATCH_SIZE);
+        Page<String> uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate,
+                pageable);
+        log.info("Found {} student participations to clean local student repositories in {} batches.", uriBatch.getTotalElements(), uriBatch.getTotalPages());
+        if (uriBatch.getTotalElements() > 0) {
+            var batchStream = Stream.iterate(uriBatch, Page::hasNext, page -> programmingExerciseStudentParticipationRepository
```
Comment (magaupp) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:138]: The `Page::hasNext` is checked before the current page is included in the Stream. The last Page has no next Page so it is not included.

**Review by magaupp — CHANGES_REQUESTED** [review]
Tested locally with 2000 students.
The Query seems to work now:
```Found 2000 student participations to clean local student repositories in 4 batches.```

However, the last batch is not processed:
```
...
Deleted Repository at repos/LOCALCOURSEEXPROGEX1/localcourseexprogex1-student1499
Deleted Repository at repos/LOCALCOURSEEXPROGEX1/localcourseexprogex1-student1500
Finished cleaning local student repositories
```

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java (hunk 7 above)
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java]: Thank you! The implementation of the iterate method and the hasNext of pages made it kinda confusing for me.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java [hunk 8]
```diff
@@ -82,49 +89,66 @@ public void cleanup() {
...
+        log.info("Search for exercises with due date from {} until {}", earliestDate, latestDate);
+        // Get all relevant participation ids
+        Pageable pageable = Pageable.ofSize(STUDENT_PARTICIPATION_CLEANUP_BATCH_SIZE);
+        Page<String> uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate,
+                pageable);
+        log.info("Found {} student participations to clean local student repositories in {} batches.", uriBatch.getTotalElements(), uriBatch.getTotalPages());
+        if (uriBatch.getTotalElements() > 0) {
+            var batchStream = Stream.iterate(uriBatch, page -> page.getNumber() <= page.getTotalPages(), page -> programmingExerciseStudentParticipationRepository
+                    .findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, page.nextPageable()));
+            batchStream.forEach(page -> page.forEach(this::deleteLocalRepositoryByUriString));
```
Comment (magaupp) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:140]: On the last page `page.nextPageable()` wraps over to `Pageable.unpaged()` which always fulfills the loop condition.
`Stream.iterate()` might not be a useful abstraction here. Try this:
```suggestion
            while (true) {
                uriBatch.forEach(this::deleteLocalRepositoryByUriString);
                if (!uriBatch.hasNext()) {
                    break;
                }
                uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, uriBatch.nextPageable());
            }
```

**Review by magaupp — CHANGES_REQUESTED** [review]
Now it's looping endlessly.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java (hunk 8 above)
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:140]: Thanks a lot. I'll check this in the next few days for other alternatives. Stephan doesn't like do while, I think while true is not the way to go and streams aren't good either in this situation. We'll see.

Comment (magaupp) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:140]: We can extract part of the first loop iteration to get an explicit loop condition:
```suggestion
            uriBatch.forEach(this::deleteLocalRepositoryByUriString);
            while (uriBatch.hasNext()) {
                uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, uriBatch.nextPageable());
                uriBatch.forEach(this::deleteLocalRepositoryByUriString);
            }
```

This, however, is not strictly DRY.

Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java:140]: I just played around it myself and actually came to the exact same conclusion as you without having it in mind.
So I think it's the best we can do. Also, I don't mind it in terms of DRY - I already extracted the deletion functionality, for the method call the parameters are different. So we have one one-liner which repeats itself once. That's fine.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java [hunk 9]
```diff
@@ -82,49 +88,69 @@ public void cleanup() {
...
+        }
+    }
+
+    private void deleteLocalRepositoryByUriString(String uri) {
+        try {
+            VcsRepositoryUri vcsRepositoryUrl = new VcsRepositoryUri(uri);
+            gitService.deleteLocalRepository(vcsRepositoryUrl);
+        }
+        catch (URISyntaxException e) {
+            log.error("Cannot create URI for repositoryUri: {} due to the following error: {}", uri, e.getMessage());
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java]: Fixed in [c2d9ea8](https://github.com/ls1intum/Artemis/pull/9322/commits/c2d9ea807cc04b6907c73807e3375b053af10c79)

**Review by magaupp — DISMISSED** [review]
Tested locally with 2000 students and 2 exercises with due date and end date set to April 1st.
Repositories are successfully deleted.

2nd cleanup run with deleted repositories for query performance measurement:
Queries finish within 300ms.
```
2024-09-28T11:58:19.271+02:00 : Cleanup git repositories on Artemis server
2024-09-28T11:58:19.271+02:00 : Search for exercises with due date from 2023-08-03T00:00+02:00[Europe/Berlin] until 2024-08-03T00:00+02:00[Europe/Berlin]
2024-09-28T11:58:19.283+02:00 : Found 4000 student participations to clean local student repositories in 8 batches.
2024-09-28T11:58:19.572+02:00 : Finished cleaning local student repositories
2024-09-28T11:58:19.572+02:00 : Search for exercises with course or exam date from 2023-08-03T00:00+02:00[Europe/Berlin] until 2024-08-03T00:00+02:00[Europe/Berlin]
2024-09-28T11:58:19.577+02:00 : Found 2 programming exercise to clean local template, test and solution: LOCALCOURSEEXPROGEX1, LOCALCOURSEEXPROGEX2
2024-09-28T11:58:19.577+02:00 : Finished cleaning local template, test and solution repositories
```

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseRepository.java [hunk 10]
```diff
@@ -229,22 +229,11 @@ default ProgrammingExercise findOneByProjectKeyOrThrow(String projectKey, boolea
...
-                LEFT JOIN FETCH pe.studentParticipations
-            WHERE pe.exerciseGroup IS NOT NULL
-                AND :endDate1 <= pe.exerciseGroup.exam.endDate
-                AND pe.exerciseGroup.exam.endDate <= :endDate2
+                LEFT JOIN FETCH pe.templateParticipation
+                LEFT JOIN FETCH pe.solutionParticipation
+            WHERE pe.id = :exerciseId
             """)
-    List<ProgrammingExercise> findAllWithStudentParticipationByRecentExamEndDate(@Param("endDate1") ZonedDateTime endDate1, @Param("endDate2") ZonedDateTime endDate2);
+    Optional<ProgrammingExercise> findWithEagerTemplateAndSolutionParticipationsById(@Param("exerciseId") long exerciseId);
```
Comment (magaupp) [inline src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseRepository.java:236]: Remove this

**Review by magaupp — CHANGES_REQUESTED** [review]
#9314 moved `findWithEagerTemplateAndSolutionParticipationsById()` from `ProgrammingExerciseRepository` into `ProgrammingExerciseTestRepository`. Your merge left the copy in `ProgrammingExerciseRepository`.

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseRepository.java (hunk 10 above)
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseRepository.java]: Thank you.

**Review by magaupp — APPROVED** [review]
re-tested

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java [hunk 11]
```diff
@@ -162,6 +164,21 @@
     Optional<ProgrammingExerciseStudentParticipation> findWithSubmissionsByExerciseIdAndStudentLoginAndTestRun(@Param("exerciseId") long exerciseId,
             @Param("username") String username, @Param("testRun") boolean testRun);
 
+    @Query("""
+            SELECT participation.repositoryUri
+            FROM ProgrammingExerciseStudentParticipation participation
+                JOIN TREAT (participation.exercise AS ProgrammingExercise) pe
+                LEFT JOIN pe.exerciseGroup eg
+                LEFT JOIN eg.exam exam
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java:172]: It's not guaranteed here that an exercise is part of an exam or exercise group.

[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java:172]: No, it's still valued. `exerciseGroup.exem` is never null, however `exercise.exerciseGroup` may be null.

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

**Review by krusche — APPROVED** [review]

**Pull request merged**